Solve basic programming model for entities (#254)
jakartaee/data: spec/src/main/asciidoc/repository.asciidoc

All Jakarta Data providers should read all declared fields directly in the class or by inheritance. However, reading fields by inheritance is not required for NoSQL databases, where support for field reading is achieved through reading directly in the entity class.

===== Basic Types

Entity data in Jakarta Data can have fields or properties of "simple" or "basic" types. These basic types are fundamental data types that Jakarta Data Providers natively support. The basic types include:

- All Java primitive types, such as int, double, boolean, etc., and their corresponding java.lang wrapper types (e.g., Integer, Double, Boolean).
- String: Represents text data.
- LocalDate, LocalDateTime, LocalTime, Instant: Represent date and time-related data.
- UUID: Universally Unique Identifier for identifying entities.
- BigInteger, BigDecimal: Represent large integer and decimal numbers.
- byte[]: Represents binary data.

> - Enum types: Custom enumerated types defined by developers.

Author: Thank you, @njr-11 let me know what you think about this.

> - Enum types: Custom enumerated types defined by developers.

Reviewer: > Thank you, @njr-11 let me know what you think about this.

The update made thus far look good.  There are still a number of unaddressed review comments, which is probably because there were a large number of comments and github hides them unless you click on the "Load more..." option.

Every entity in Jakarta Data must have a unique identifier composed of one or more supported basic types. This unique identifier is crucial for distinguishing individual entities in the database. Developers should define and annotate the identifier fields in their entity classes to ensure proper identification and retrieval of entities.

> IMPORTANT: It's important to note that key-value, wide-column, document, and relational databases support Collection specializations and Maps of the basic types. However, these databases may have different serialization processes, impacting performance and impedance mismatch. Developers should consider these considerations when working with collections and maps in their entity models.

Reviewer: ```suggestion
IMPORTANT: It's important to note that key-value, wide-column, document, and relational databases support Collection specializations and Maps of the basic types. However, these databases may have different serialization processes, impacting performance and causing impedance mismatch. Developers should consider these side effects when working with collections and maps in their entity models.
```

> In addition to the basic types, Jakarta Data Providers should enable support for custom types. Custom types are user-defined classes that represent data in a domain-specific manner. These custom types may include complex data structures and objects. Jakarta Data Providers should provide a mechanism to convert these custom types to the supported basic types.

Reviewer: ```suggestion
In addition to the basic types, entity models might also choose to support additional types that represent data in a domain-specific manner. These custom types may include complex data structures and objects. Entity models can choose to provide mechanisms to convert these custom types to the supported basic types.
```

> 
> One way to achieve this is by using the `Converter` annotation, similar to the one found in Jakarta Persistence (JPA). The `Converter` annotation allows developers to specify how custom types should be converted to and from the basic types. It ensures that custom data structures can be seamlessly integrated into the entity model.

Reviewer: Jakarta hasn't yet defined how the user supplies a Converter to a Jakarta Data provider.  Let's leave this part out until we define it.
```suggestion
```

===== Domain-Relation Fields in Jakarta Data

In Jakarta Data, the concept of "Domain-Relation" fields encompasses two distinct types: "component" and "association" fields. These fields enable developers to establish relationships between entities and other domain concepts, enriching the complexity and structure of data entities.

- *Component Fields*: A component field represents a relationship where one entity is treated as a component of another entity. It implies that the component does not have a life cycle outside the entity to which it belongs. It is an embedded object that exists solely within the context of the owning entity.

- *Association Fields*: Association fields represent semantically weak relationships, often called semantic dependencies, between objects that may not have a direct or strong connection. Unlike component fields, association fields may exist between entities that are otherwise unrelated. Associations can further specialize into aggregation, a specific form of association in which each object involved has its life cycle, yet a notion of ownership also exists.

> These Domain-Relation fields' specific implementation and definition may vary across different Jakarta Data providers. Providers can create annotations, define conventions, or leverage standards like Jakarta Persistence to establish these relationships effectively.

Reviewer: ```suggestion
Definition and implementation of Domain-Relation fields may vary across different Jakarta Data providers. Providers can create annotations, define conventions, or leverage standards like Jakarta Persistence to establish these relationships effectively.
```

> The topic of serialization of these Domain-Relation fields within databases is crucial for maintaining data consistency and integrity. In the next section, we'll delve into how Jakarta Data providers handle the persistence and serialization of these fields within various database types.

Reviewer: ```suggestion
The topic of serialization of Domain-Relation fields within databases is crucial for maintaining data consistency and integrity. The next section discussed how Jakarta Data providers handle the persistence and serialization of Domain-Relation fields for various database types.
```

In Jakarta Data, the serialization of Domain-Relation fields, such as components, can be achieved in two ways.

1. *Merging Fields Directly in the Entity (Component Embedding):*

In this approach, a component merges its fields directly within the entity. From the perspective of the persistence layer, the component itself doesn't exist as a separate table or document. Instead, it becomes part of the entity's structure. This approach leads to a flat representation in the database.

Example:

[source,java]
----
public class Address {

    private String street;
    private String city;
    private String postalCode;
}

//the entity
public class Person {
    private Long id;

    private String name;
    private Address address;  // This is a component field
}
----

> The structure of a Document, wide-column, and Graph will be like the JSON representation below. The JSON representation is for illustration; it belongs to the NoSQL database and defines the final serialization structure such as JSON and BSON, including another field such as timestamp for clock-vector, revision, and other internal purposes. The JSON and the following SQL table are representations resulting from the Java classes.

Reviewer: ```suggestion
The structure of the entity in a Document, Wide-Column, and Graph database will be like the JSON representation below. The JSON representation is for illustration; the actual representation is an implementation detail of the database. The JSON and the following SQL table are representations resulting from the Java classes.
```

[source,json]
----
{
  "id": 1,
  "name": "John Doe",
  "street": "123 Main St",
  "city": "Sampleville",
  "postalCode": "12345"
}
----

[options="header"]
|====

> | id | name     | street        | city
> | 1  | John Doe | 123 Main St   | Sampleville

Reviewer: ```suggestion
| id | name     | street        | city         | postalCode
| 1  | John Doe | 123 Main St   | Sampleville  | 12345
```

|====

> This approach allows for a flat and denormalized structure in the database, making it suitable for certain NoSQL database types.

Reviewer: ```suggestion
This approach allows for a flat and denormalized structure in the database, making it suitable for a variety of database types.
```

For instance, consider the `Citizen` and `Passport` classes. In a relational database, this approach results in two separate tables, each representing an entity with its associated persistence context. In contrast, for a NoSQL database like a Document database, the `Passport` can still act as a component within the `Citizen` entity, signifying that a `Passport` is closely tied to a `Citizen`. However, the modeling may vary depending on the specific NoSQL database and its capabilities.

[source,java]
----
public class Passport {
    private Long id;
    private String passportNumber;
    private LocalDate expirationDate;
}

public class Citizen {
    private Long id;
    private String name;
    private Passport passport; // One-to-One relationship with Passport for relational database
}
----

> Here's a representation of the `Citizen` and `Passport` classes in both table and JSON formats:

Reviewer: ```suggestion
Here are some possible representations of the `Citizen` and `Passport` entities in table and JSON formats:
```

[options="header"]
|===
| Citizen Table | Passport Table
| id (Primary Key) | id (Primary Key)
| name | passportNumber
| passport_id (Foreign Key) | expirationDate

> |===

Reviewer: It is confusing to show 2 database tables in a single table.  I would expect to see two tables for this, one for each, and each including the same example data that you show for the JSON example.

[source,java]
----
// Entity
public class Author {

    private UUID id;

    private String name;

    private List<Book> books;
}

// If in a relational database, Book might also be an entity
public class Book {

    // This field might not be required for some NoSQL database modeling
    private Long id;

    private String title;

    private String category;

    // Other fields and methods
}
----

[source,json]
----
{
  "id": "6f6d665d-5585-46cd-8b9b-61a559de0e13",
  "name": "John Doe",
  "books": [
    {
      "title": "Sample Book 1",
      "category": "Fiction"
    },
    {
      "title": "Sample Book 2",
      "category": "Non-fiction"
    }
  ]
}
----

In the JSON representation, an `Author` entity can be associated with multiple `Book` entities within an array.

[options="header"]
|===
| Author Table | Book Table | Author_Book (Auxiliary) Table

| id (Primary Key) | id (Primary Key) | author_id (Foreign Key)

> | name | title | book_id (Foreign Key)

Reviewer: This should show each table separately along with the same data from the JSON example.

|===

In a relational database, this relationship typically generates three tables: `Author`, `Book`, and an auxiliary table `Author_Book` to manage the many-to-many relationship between authors and books.

In some scenarios, books can have multiple authors, and authors can contribute to several books, resulting in a many-to-many (N-N) cardinality relationship. Jakarta Data offers flexibility in representing and managing such complex relationships.

This N-N relationship typically generates a dedicated table to manage the associations between books and authors in a relational database. However, NoSQL databases may take a different approach, especially in cases where denormalization and data duplication are favored for query-driven designs.

> IMPORTANT: The key-value database might support these fields, generating a single BLOB value. However, the serialization process for such fields may vary depending on the Jakarta Data provider and the specific NoSQL database used.

Reviewer: ```suggestion
IMPORTANT: The key-value database might support these fields, generating a single BLOB value. However, the serialization process for such fields may vary depending on the Jakarta Data provider and the specific key-value database used.
```

> NOTE: The Graph database is not required to support this type, but it might be used for aggregate query returns or as a read-only field.

Reviewer: ```suggestion
NOTE: The Graph database is not required to support Domain-Relation types, but it might be used for aggregate query returns or as a read-only field.
```

===== Recursion at Domain-Relation Fields in Jakarta Data

In the context of Jakarta Data, the term "recursion" pertains to the ability to manage hierarchical or nested relationships between entities. This capability is essential when dealing with complex domain models involving associations, aggregations, or compositions between various entity types. Jakarta Data ensures these relationships are correctly mapped and maintained within the database, enabling consistent data retrieval and manipulation.

> Within relational databases, Jakarta Data necessitates support for recursive relationships. It implies that if one entity type is associated with or contains another entity type, the Jakarta Data provider for relational databases must establish the required table structures and foreign key constraints to uphold these relationships. This approach guarantees data integrity and consistency when working with the database.

Reviewer: ```suggestion
For relational databases, Jakarta Data requires support of recursive relationships. If one entity type is associated with or contains another entity type, the Jakarta Data provider for relational databases must establish the required table structures and foreign key constraints to uphold these relationships. This approach guarantees data integrity and consistency when working with the database.
```

> On the other hand, NoSQL databases, known for their flexible and schema-less nature, may not require explicit support for recursive relationships as dictated by the Jakarta Data specification. In NoSQL databases, data is often stored in a denormalized or nested document format, making it more challenging to enforce strict hierarchical relationships. Instead, NoSQL databases may emphasize query-driven design and may not necessitate explicit mapping or management of recursive relationships.

Reviewer: ```suggestion
For other types of databases, Jakarta Data does not require explicit support for recursive relationships.  In NoSQL databases, data is often stored in a denormalized or nested document format, making it more challenging to enforce strict hierarchical relationships. Instead, NoSQL databases may emphasize query-driven design rather than explicit mapping or management of recursive relationships.
```

> In instances where a Jakarta Data provider for NoSQL databases encounters a recursive relationship that it cannot support due to the specific characteristics of the database, it should throw a `jakarta.data.exceptions.MappingException` or an appropriate specialized exception class. This exception is notified to developers that the Jakarta Data provider does not support this specific use case within the NoSQL database environment.

Reviewer: ```suggestion
In instances where a Jakarta Data provider for NoSQL databases encounters a recursive relationship that it cannot support due to the specific characteristics of the database, it must throw a `jakarta.data.exceptions.MappingException` or an appropriate subclass of `MappingException`. This exception notifies developers that the database does not support the relationship.
```

[source,java]
----
// Entity
public class Author {

    private UUID id;

    private String name;

    private List<Book> books;
}

// If in a relational database, Book might also be an entity
public class Book {

    // This field might not be required for some NoSQL database modeling
    private Long id;

    private String title;

    private String category;

    // Other fields and methods
}
----

[source,json]
----
{
  "id": "6f6d665d-5585-46cd-8b9b-61a559de0e13",
  "name": "John Doe",
  "books": [
    {
      "title": "Sample Book 1",
      "category": "Fiction"
    },
    {
      "title": "Sample Book 2",
      "category": "Non-fiction"
    }
  ]
}
----

In the JSON representation, an `Author` entity can be associated with multiple `Book` entities within an array.

> Certainly, here are the tables with content based on the JSON data structure:

Reviewer: ```suggestion
Here are the tables with content based on the JSON data structure:
```

.Author Table:
[options="header"]
|===
| id (Primary Key) | name
| 1                | John Doe
| 2                | Jane Smith
|===

.Book Table:
[options="header"]
|===
| id (Primary Key) | title | category

> | 1                | Book A | Fiction
> | 2                | Book B | Non-fiction

Reviewer: Data values should be the same in both JSON and table examples given that we previously stated the table contents are based on the JSON data structure,
```suggestion
| 1                | Sample Book 1 | Fiction
| 2                | Sample Book 2 | Non-fiction
```
